data/code/lib.c

```diff
@@ -589,3 +593,732 @@ void check_lap_ex()
+{
+    if (!is_touching_ground(this))
+    {
+        return false;
+    }
+    return *(int*)(this + KART_BODY_CURRENT_TERRAIN_TYPE_OFFSET) == EXTENDED_TERRAIN_BOUNCY;
+}
+
+// Moves Kart position directly. Bad when done in large amounts, which is why XZ movement is small.
+void add_absolute_position_vector(char* const this,
+                                  float movement_vector_x,
+                                  float movement_vector_y,
+                                  float movement_vector_z)
+{
+    float* vector_x_value = (float*)(this + KART_BODY_POS_X_OFFSET);
+    float* vector_y_value = (float*)(this + KART_BODY_POS_Y_OFFSET);
+    float* vector_z_value = (float*)(this + KART_BODY_POS_Z_OFFSET);
+
+    *vector_x_value += movement_vector_x;
+
+    *vector_y_value += movement_vector_y;
+
+    *vector_z_value += movement_vector_z;
+}
+
+// Add to Kart's movement vector. Used for Y adjustment during bounce.
+void add_movement_vector(char* const this,
+                         float movement_vector_x,
+                         float movement_vector_y,
+                         float movement_vector_z)
+{
+    float* vector_x_value = (float*)(this + KART_BODY_MOV_VECTOR_X_OFFSET);
+    float* vector_y_value = (float*)(this + KART_BODY_MOV_VECTOR_Y_OFFSET);
+    float* vector_z_value = (float*)(this + KART_BODY_MOV_VECTOR_Z_OFFSET);
+
+    *vector_x_value += movement_vector_x;
+
+    *vector_y_value += movement_vector_y;
+
+    *vector_z_value += movement_vector_z;
+}
+
+// Overwrite the Kart's movement vector. Used during liftoff.
+void write_movement_vector(char* const this,
+                           float movement_vector_x,
+                           float movement_vector_y,
+                           float movement_vector_z)
+{
+    float* vector_x_value = (float*)(this + KART_BODY_MOV_VECTOR_X_OFFSET);
+    float* vector_y_value = (float*)(this + KART_BODY_MOV_VECTOR_Y_OFFSET);
+    float* vector_z_value = (float*)(this + KART_BODY_MOV_VECTOR_Z_OFFSET);
+
+    *vector_x_value = movement_vector_x;
+
+    *vector_y_value = movement_vector_y;
+
+    *vector_z_value = movement_vector_z;
+}
+
+// Stops kart from falling too fast. Must do manually as DoSpeedCtrl is hijacked and not running.
+void clamp_movement_vector_descent(char* const this)
+{
+    float* movement_vector_y = (float*)(this + KART_BODY_MOV_VECTOR_Y_OFFSET);
+    if (*movement_vector_y <= BOUNCE_DESCENT_CAP)
+    {
+        float* movement_vector_x = (float*)(this + KART_BODY_MOV_VECTOR_X_OFFSET);
+        float* movement_vector_z = (float*)(this + KART_BODY_MOV_VECTOR_Z_OFFSET);
+        write_movement_vector(this, *movement_vector_x, BOUNCE_DESCENT_CAP, *movement_vector_z);
+    }
+}
+
+// Gets 4-bytle hash from ground traingle material.
+int get_ground_hash(char* const this)
+{
+    int* pointer = (int*)(this + KART_BODY_COLLISION_TRIANGLE_POINTER_OFFSET);
+    int* pointerpointer = (int*)(*pointer + CRS_GROUND_COLLISION_TRIANGLE_POINTER_OFFSET);
```

Comment:
I guess `pointerpointer` is a pointer to `CollisionTriangle` and `pointer` to `CrsGround`?

Reply:
I've been helping others with projects and I don't think I'd go back to this method of fetching object data now.